snow/engine/snowman/bootstrap/bootstrapper.go

```diff
@@ -566,168 +555,128 @@ func (b *Bootstrapper) markUnavailable(nodeID ids.NodeID) {
-//     VM to have stored blocks during `ParseBlock`.
-//
-// If [blk]'s height is <= the last accepted height, then it will be removed
-// from the missingIDs set.
-func (b *Bootstrapper) process(ctx context.Context, blk snowman.Block, processingBlocks map[ids.ID]snowman.Block) error {
-	for {
-		blkID := blk.ID()
-		if b.Halted() {
-			// We must add in [blkID] to the set of missing IDs so that we are
-			// guaranteed to continue processing from this state when the
-			// bootstrapper is restarted.
-			b.Blocked.AddMissingID(blkID)
-			return b.Blocked.Commit()
-		}
-
-		b.Blocked.RemoveMissingID(blkID)
-
-		status := blk.Status()
-		// The status should never be rejected here - but we check to fail as
-		// quickly as possible
-		if status == choices.Rejected {
-			return fmt.Errorf("bootstrapping wants to accept %s, however it was previously rejected", blkID)
-		}
-
-		blkHeight := blk.Height()
-		if status == choices.Accepted || blkHeight <= b.startingHeight {
-			// We can stop traversing, as we have reached the accepted frontier
-			if err := b.Blocked.Commit(); err != nil {
-				return err
-			}
-			return b.tryStartExecuting(ctx)
-		}
+//   - ancestors is a set of blocks that can be used to optimisically lookup
+//     parent blocks. This enables the engine to process multiple blocks without
+//     relying on the VM to have stored blocks during `ParseBlock`.
+func (b *Bootstrapper) process(
+	ctx context.Context,
+	blk snowman.Block,
+	ancestors map[ids.ID]snowman.Block,
+) error {
+	lastAcceptedHeight, err := b.getLastAcceptedHeight(ctx)
+	if err != nil {
+		return err
+	}
 
-		// If this block is going to be accepted, make sure to update the
-		// tipHeight for logging
-		if blkHeight > b.tipHeight {
-			b.tipHeight = blkHeight
-		}
+	numPreviouslyFetched := b.tree.Len()
 
-		pushed, err := b.Blocked.Push(ctx, &blockJob{
-			log:         b.Ctx.Log,
-			numAccepted: b.numAccepted,
-			blk:         blk,
-			vm:          b.VM,
-		})
-		if err != nil {
-			return err
-		}
+	batch := b.DB.NewBatch()
+	missingBlockID, foundNewMissingID, err := process(
+		batch,
+		b.tree,
+		b.missingBlockIDs,
+		lastAcceptedHeight,
+		blk,
+		ancestors,
+	)
+	if err != nil {
+		return err
+	}
 
-		if !pushed {
-			// We can stop traversing, as we have reached a block that we
-			// previously pushed onto the jobs queue
-			if err := b.Blocked.Commit(); err != nil {
-				return err
-			}
-			return b.tryStartExecuting(ctx)
-		}
+	// Update metrics and log statuses
+	{
+		numFetched := b.tree.Len()
+		b.numFetched.Add(float64(b.tree.Len() - numPreviouslyFetched))
 
-		// We added a new block to the queue, so track that it was fetched
-		b.numFetched.Inc()
+		height := blk.Height()
+		b.tipHeight = max(b.tipHeight, height)
 
-		// Periodically log progress
-		blocksFetchedSoFar := b.Blocked.Jobs.PendingJobs()
-		if blocksFetchedSoFar%statusUpdateFrequency == 0 {
+		if numFetched%statusUpdateFrequency == 0 {
```

Comment:
Because `numFetched` is now increased in batches, this check doesn't work correctly anymore.